File: valid-palindrome.cs
using System.Text;

public class Solution
{
    static private string PreProcessing(string s)
    {
        StringBuilder builder = new StringBuilder();
        foreach (char c in s)
            if (char.IsLetterOrDigit(c))
                builder.Append(char.ToLower(c));
        return builder.ToString();
    }

    public bool IsPalindrome(string s)
    {
        if (s == null || s.Length == 0)
            return true;
        string Temp = PreProcessing(s);
        int i = 0, j = Temp.Length - 1;
        while (i < j)
        {
            if (Temp[j--] != Temp[i++])
                return false;
        }
        return true;
    }
}

File: valid-palindrome.cc
#include <string>
#include <cctype>

class Solution
{
public:
    void preprocess(std::string &s)
    {
        std::string temp;
        for (const auto c : s)
            if (std::isalnum(c))
                temp.push_back(std::tolower(c));
        s.swap(temp);
    }

    bool isPalindrome(std::string &s)
    {
        preprocess(s);
        int l = 0, r = s.size() - 1;
        while (l < r)
            if (s[l++] != s[r--])
                return false;
        return true;
    }
};